base/iobuffer.jl

```diff
@@ -322,113 +483,142 @@ function seekend(io::GenericIOBuffer)
-            if (io.size-io.offset+nshort > io.maxsize) ||
-                (datastart > 4096 && datastart > io.size - io.ptr) ||
-                (datastart > 262144)
-                # apply somewhat arbitrary heuristics to decide when to destroy
-                # old, read data to make more room for new data
-                compact(io)
-            end
+    reclaimable_bytes = first(get_used_span(io)) - 1
+    available_bytes = min(lastindex(io.data), io.maxsize) - (io.append ? io.size : io.ptr - 1)
+    # Avoid resizing and instead compact the buffer, only if we gain enough bytes from
+    # doing so (at least 32 bytes and 1/8th of the data length). Also, if we would have
+    # to resize anyway, there would be no point in compacting, so also check that.
+    if (
+            reclaimable_bytes ≥ 32 &&
+            reclaimable_bytes ≥ length(io.data) >>> 3 &&
+            (reclaimable_bytes + available_bytes) % UInt ≥ nshort
+        )
+        compact!(io)
+        return io
+    end
+
+    desired_size = length(io.data) + Int(nshort) - available_bytes
+    if desired_size > io.maxsize
+        # If we can't fit all the requested data in the new buffer, we need to
+        # fit as much as possible, so we must compact
+        if !iszero(reclaimable_bytes)
+            desired_size -= compact!(io)
         end
+        # Max out the buffer size if we want more than the buffer size
+        if length(io.data) < io.maxsize
+            _resize!(io, io.maxsize, true)
+        end
+    else
+        # Else, we request only the requested size, but set `exact` to `false`,
+        # in order to overallocate to avoid growing the buffer by too little
+        _resize!(io, desired_size, false)
     end
-    return
+
+    return io
 end
 
-@inline ensureroom(io::GenericIOBuffer, nshort::Int) = ensureroom(io, UInt(nshort))
-@inline function ensureroom(io::GenericIOBuffer, nshort::UInt)
-    if !io.writable || (!io.seekable && io.ptr > io.offset+1) || io.reinit
-        ensureroom_slowpath(io, nshort)
-    end
-    n = min((nshort % Int) + (io.append ? io.size : io.ptr-1) - io.offset, io.maxsize)
-    l = length(io.data) + io.offset
-    if n > l
-        _resize!(io, Int(n))
+# Get the indices in data which cannot be deleted
+function get_used_span(io::IOBuffer)
+    # A seekable buffer can recover data before ptr
+    return if io.seekable
+        io.offset + 1 : io.size
+    # If non-seekable, the mark can be used to recover data before ptr,
+    # so data at the mark and after must also be saved
+    elseif io.mark > -1
+        min(io.ptr, io.mark + io.offset + 1) : io.size
+    else
+        io.ptr : io.size
     end
-    return io
+end
+
+# Return number of bytes recovered
+function compact!(io::GenericIOBuffer)::Int
+    used_span = get_used_span(io)
+    offset = first(used_span) - 1
+    iszero(offset) && return 0
+    data = io.data
+    copyto!(data, 1, data, offset + 1, length(used_span))
+    io.offset = 0
```

Comment:
```suggestion
    io.offset -= offset # preserves mark/position computations after compacting shifted the data
    ismarked(io) || (io.offset = 0) # but not required to preserve position/mark computation if the stream was not marked (this step is optional)
```

(plus some changes to `_resize!` / `ensureroom_slowpath` since directly adding `offset + maxsize` is not correct anymore, but rather likely needs to use `min(io.maxsize, length(get_used_span) + n_requested_short)` instead of `min(lastindex(io.data), io.maxsize)`)

Reply:
This is handled in [819d750](https://github.com/JuliaLang/julia/pull/57570/commits/819d7506b51d385e534d043042567750687cf621) now.